File: app/src/main/java/netpro/keyTransmitter/EditKeyRecyclerViewAdapter.java
package netpro.keyTransmitter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import java.util.LinkedList;
import java.util.List;

public class EditKeyRecyclerViewAdapter extends RecyclerView.Adapter<EditKeyViewHolder> {
    private List<Key> datasource = new LinkedList<>();
    private int displayWidth = 720;
    //private int width = 768;
    private Context context;
    private OnRecyclerClickListener listener;

    public EditKeyRecyclerViewAdapter(Context context) {
        this.context = context;
    }

    @Override
    public EditKeyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_cardview, parent, false);
        return new EditKeyViewHolder(v);
    }

    @Override
    public void onBindViewHolder(EditKeyViewHolder view, int position) {
        //Keyのサイズを画面のサイズに合わせて変更
        Key key = datasource.get(position);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        params.width = (displayWidth - displayWidth / 8) / 4 * key.getColumnSpan();
        params.height = (displayWidth - displayWidth / 8) / 4 * key.getRowSpan();

        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        if (key.getColumnSpan() > 1) {
            params.width += (key.getColumnSpan() - 1) * (10 * metrics.density);
        }
        if (key.getRowSpan() > 1) {
            params.height += (key.getRowSpan() - 1) * (10 * metrics.density);
        }

        view.itemView.setLayoutParams(params);
        view.setKey(key);
        view.setListener(listener);
    }

    @Override
    public int getItemCount() {
        return datasource.size();
    }

    public void addView(Key key) {
        datasource.add(key);
        notifyItemInserted(getItemCount() - 1);
    }

    public void addAllView(List<Key> list) {
        int preSize = getItemCount();
        datasource.addAll(list);
        notifyItemRangeInserted(preSize, list.size() - 1);
    }

    public void removeView(int index) {
        Key target = datasource.get(index);
        int keySize = target.getColumnSpan() * target.getRowSpan();
        datasource.remove(index);
        notifyItemRemoved(index);
        for (int i = 0; i < keySize; i++) {
            addView(new EmptyKey());
            notifyItemInserted(datasource.size() - 1);
        }
        Log.d("remove", String.valueOf(index));
    }

    public void move(int fromIndex, int toIndex) {
        notifyItemMoved(fromIndex, toIndex);
        Key temp = datasource.get(fromIndex);
        datasource.remove(fromIndex);
        datasource.add(toIndex, temp);
    }

    public Key get(int index) {
        return datasource.get(index);
    }

    public List<Key> getDatasource() {
        return datasource;
    }

    public void setDatasource(List<Key> datasource) {
        this.datasource = datasource;
        notifyDataSetChanged();
    }

    public String getName() {
        return "keyboard";
    }

    public void setOnRecyclerClickListener(OnRecyclerClickListener listener) {
        this.listener = listener;
    }
}

File: app/src/main/java/netpro/keytransmitter/EditKeyRecyclerViewAdapter.kt
package netpro.keytransmitter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

/**
 * EditActivityのRecyclerViewのアダプタ
 */
class EditKeyRecyclerViewAdapter : RecyclerView.Adapter<EditKeyViewHolder>() {
    var keyList: MutableList<Key> = mutableListOf()

    private lateinit var listener: OnRecyclerClickListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EditKeyViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.layout_cardview, parent, false)
        return EditKeyViewHolder(v)
    }

    override fun onBindViewHolder(view: EditKeyViewHolder, position: Int) {
        val key = keyList[position]
        view.setKey(key)
        view.setListener(listener)
    }

    override fun getItemCount(): Int {
        return keyList.size
    }

    fun addView(key: Key) {
        keyList.add(key)
        notifyItemInserted(itemCount - 1)
    }

    fun addView(index: Int, key: Key) {
        keyList.add(index, key)
        notifyItemInserted(index)
    }

    fun addAllView(list: List<Key>) {
        val preSize = itemCount
        keyList.addAll(list)
        notifyItemRangeInserted(preSize, list.size - 1)
    }

    fun removeView(index: Int, fillsEmpty: Boolean) {
        val target = keyList[index]
        keyList.removeAt(index)
        notifyItemRemoved(index)

        if (!fillsEmpty) {
            return
        }

        val keySize = target.columnSpan * target.rowSpan
        for (i in 0..keySize - 1) {
            addView(EmptyKey())
            notifyItemInserted(keyList.size - 1)
        }
    }

    fun move(fromIndex: Int, toIndex: Int) {
        notifyItemMoved(fromIndex, toIndex)
        val temp = keyList[fromIndex]
        keyList.removeAt(fromIndex)
        keyList.add(toIndex, temp)
    }

    operator fun get(index: Int): Key {
        return keyList[index]
    }

    fun setOnRecyclerClickListener(listener: OnRecyclerClickListener) {
        this.listener = listener
    }

    val emptySpace: Int
        get() {
            return keyList
                    .filterIsInstance<EmptyKey>()
                    .sumBy { it.columnSpan * it.rowSpan }
        }
}